README.md
# @o-rango/update-notify

[![Build Status](https://travis-ci.com/o-rango/update-notify.svg?branch=master)](https://travis-ci.com/o-rango/update-notify) 
[![codecov](https://codecov.io/gh/o-rango/update-notify/branch/master/graph/badge.svg)](https://codecov.io/gh/o-rango/update-notify)



Inspired in update-notifier but using the default `npm` api that way is working behind `proxy` or any other config reflected in `.npmrc`.

It only has programmatic api and can be easily attached to your workflow.

## Install 

```
npm i --save @o-rango/updatet-notify
```


## Usage

```js
const { updateNotify } = require('@o-rango/update-notify');

const notifier = updateNotify({
      pkg: {
        version: '0.0.1',
        name: 'react'
      },
      updateCheckInterval: 0
    });


console.log(notifier);
// Output =>  
```


## API

### notify = updateNotify(options)

Checks if there is an available update. Accepts options defined below. Returns `{ updateAvailable: true , name: 'react' , latest: '16.11.0', 'current': '0.0.1' }` if there is any update.

### options

Type: `object`

#### pkg

Type: `object`

##### name

*Required*<br>
Type: `string`

##### version

*Required*<br>
Type: `string`

#### updateCheckInterval

Type: `number`<br>
Default: `1000 * 60 * 60 * 24` *(1 day)*

How often to check for updates.

#### distTag

Type: `string`<br>
Default: `latest`

Which [dist-tag](https://docs.npmjs.com/adding-dist-tags-to-packages) to use to find the latest version.


## Extra 

### ping = pingRegistry()

Easter egg method that pings registry defined in your `.npmrc`;


```js
const { pingRegistry } = require('@o-rango/update-notify');


const online = await pingRegistry();
// Output online => :{ isOnline : true }

```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.1](https://github.com/o-rango/update-notify/compare/v0.1.0...v0.1.1) (2020-01-22)


### Bug Fixes

* **deps:** update dependency execa to v3.4.0 ([#6](https://github.com/o-rango/update-notify/issues/6)) ([a7b0a83](https://github.com/o-rango/update-notify/commit/a7b0a83948a058b4b69a63f4064ba628455f64ad))
* **deps:** update dependency execa to v4 ([#9](https://github.com/o-rango/update-notify/issues/9)) ([7feaba8](https://github.com/o-rango/update-notify/commit/7feaba8aa9c635301247146400f5401261783c79))
* **deps:** update dependency semver to v7 ([#8](https://github.com/o-rango/update-notify/issues/8)) ([4d74c5e](https://github.com/o-rango/update-notify/commit/4d74c5e5512220fccdd61d50d1d1d4a460b507da))

## [0.1.0](https://github.com/o-rango/update-notify/compare/v0.0.4...v0.1.0) (2019-12-19)


### ⚠ BREAKING CHANGES

* n

### Bug Fixes

* 🐛 auto-update not writting new configstore ([5499180](https://github.com/o-rango/update-notify/commit/54991805039f5f11afa5a72dd2f46ccb282c3707))

### [0.0.4](https://github.com/o-rango/update-notify/compare/v0.0.3...v0.0.4) (2019-11-18)

### [0.0.3](https://github.com/o-rango/update-notify/compare/v0.0.2...v0.0.3) (2019-11-18)


### Bug Fixes

* 🐛 hangs on ping proxied registry ([60bfbf9](https://github.com/o-rango/update-notify/commit/60bfbf9c73fe656cf9aa13220b2782972fd2fe86))

### 0.0.2 (2019-11-17)


### Features

* initial commit ([ab4a2c9](https://github.com/o-rango/update-notify/commit/ab4a2c9c9fc73cfb12ba23a98a75d9e6d4ebfd77))


test/helpers.test.js
const {
  semverCheck,
  pingRegistry,
  latestVersion,
  setConfig,
  getConfig,
  shouldCheckUpdates,
  notifyFlow
} = require('../src/helpers');

describe('Helpers Test - Semver', () => {
  test('Semver check ', async () => {
    const result = semverCheck('0.0.1', '0.0.2');
    expect(result).toBe(true);
  });

  test('Semver check ', async () => {
    const result = semverCheck('0.1.1', '0.0.2');
    expect(result).toBe(false);
  });
});

describe('Helpers Test - pingRegistry', () => {
  test('pingRegistry', async () => {
    const result = pingRegistry();
    await expect(result).resolves.toHaveProperty('isOnline', true);
  });
});

describe('Helpers Test - latestVersion', () => {
  test('latestVersion', async () => {
    const result = latestVersion('react', 'latest');
    await expect(result).resolves.toContain('16');
  });
});

describe('Helpers Test - Configs', () => {
  test('setConfig', async () => {
    const result = setConfig('react');
    expect(typeof result.path).toBe('string');
  });

  xtest('getConfig', async () => {
    const result = getConfig('react', 'lastUpdateCheck');
    expect(result).toBeTruthy();
  });
});

describe('Helpers Test - Should Check for updates', () => {
  test('shouldCheckUpdates - no', async () => {
    const result = shouldCheckUpdates(Date.now(), 1000 * 60 * 60 * 5);
    expect(result).toBe(false);
  });

  test('shouldCheckUpdates - yes ', async () => {
    const result = shouldCheckUpdates(1574012350978, 0);
    expect(result).toBe(true);
  });

  test('no data should fail - no ', async () => {
    const result = shouldCheckUpdates();
    expect(result).toBe(false);
  });
});

describe('Helpers Test - NotifyFlow', () => {
  test('notifyFlow - should update', async () => {
    // Never Returns
    const result = await notifyFlow({
      pkg: { name: 'react', version: '0.10.1' }
    });
    await expect(result).toBeUndefined();
  });
});


src/helpers.js
'use strict';
const importLazy = require('import-lazy')(require);
const ciInfo = importLazy('ci-info');
const execa = importLazy('execa');
const semver = importLazy('semver');
const ConfigStore = importLazy('configstore');

/**
 *  @method isCI Check if is running on CI ENVIRONMENT
 * @returns {boolean} is CI or is not CI
 */
const isCI = function() {
  return ciInfo.isCI;
};

/**
 * @method pingRegistry
 * Check if there is connectivity with registry having short-circuit 5 seconds fail
 */
const pingRegistry = async function() {
  const timer = setTimeout(process.exit, 1000 * 5);
  const { failed } = await execa.command('npm ping', { detached: false });

  const exit = () => {
    clearTimeout(timer);
    return !failed;
  };

  return { isOnline: (await failed) ? failed : exit() };
};

/**
 * @method latestVersion checks latest version of pkg in registry.
 * @param {string} pkgName package name
 * @param {string} distTag type of tags to search against registry
 * @returns {Object} latest version of package in registry
 */
const latestVersion = async function(pkgName, distTag = 'latest') {
  try {
    const { stdout, stderr } = await execa.command(
      `npm info ${pkgName} dist-tags.${distTag}`,
      {
        detached: process.platform !== 'win32'
      }
    );
    if (stderr !== '') return;
    return await stdout;
  } catch (error) {
    return;
  }
};

/**
 * @method pkgSemver
 * @description compare if packages satisfies semver
 */
const semverCheck = function(currentVersion, latestVersion) {
  return semver.lt(currentVersion, latestVersion);
};

/**
 * @method setConfig runs configstore and save configuration
 * @param packageName Some name to say hello for.
 * @param params extra configuration to save.
 */
const setConfig = function(packageName, params = {}) {
  try {
    const result = new ConfigStore(`o-update-notify-${packageName}`);
    result.set(params);
    return result;
  } catch (error) {
    // Expecting error code EACCES or EPERM
    console.error(`
    ${error} , Looks like EACCES or EPERM error ocurred, Try running with %s or get access sudo permissions.
  `);
  }
};

/**
 * @method getConfig runs configstore and save configuration
 * @param packageName key to retrieve config.
 * @param config  value to be retrieved from the config.
 */
const getConfig = function(packageName, config = '') {
  const result = new ConfigStore(`o-update-notify-${packageName}`).get(config);
  return result;
};

/**
 * @method shouldCheckUpdates runs configstore and save configuration
 * @param lastUpdateCheck when notify-update was last checked
 * @param updateCheckInterval interval that configuration must be checked
 * @returns {boolean} if it should run or not
 */
const shouldCheckUpdates = function(lastUpdateCheck, updateCheckInterval) {
  if (!lastUpdateCheck) return false;
  return !(Date.now() - lastUpdateCheck < updateCheckInterval);
};

/**
 * @method notifyFlow workflow of check and save configuration
 * @param options options to be passed over the fn
 */
const notifyFlow = async function(options) {
  const { name, version } = options.pkg || {};
  const { distTag } = options.distTag || 'latest';
  const lVersion = await latestVersion(name, distTag);
  if (!lVersion) return;
  const updateResult = semverCheck(version, lVersion);
  setConfig(name, {
    updateAvailable: updateResult,
    latest: lVersion,
    current: version,
    lastUpdateCheck: Date.now()
  });
};

module.exports = {
  isCI,
  pingRegistry,
  latestVersion,
  semverCheck,
  setConfig,
  getConfig,
  shouldCheckUpdates,
  notifyFlow
};


src/index.js
const {
  isCI,
  pingRegistry,
  semverCheck,
  getConfig,
  shouldCheckUpdates,
  notifyFlow
} = require('./helpers');

const updateNotify = function(options = {}) {
  // Never run on CI
  if (isCI()) return;
  if (!options.pkg.name || !options.pkg.version)
    throw new Error('pkg.name and pkg.version required');
  const { updateCheckInterval = 1000 * 60 * 60 * 24 } = options;
  const lastUpdateCheck = getConfig(options.pkg.name, 'lastUpdateCheck');
  const latestCachedVersion = getConfig(options.pkg.name, 'latest');
  const updateAvailable = getConfig(options.pkg.name, 'updateAvailable');

  if (!lastUpdateCheck && !latestCachedVersion && !updateAvailable)
    notifyFlow(options); // 1st Run
  if (!shouldCheckUpdates(lastUpdateCheck, updateCheckInterval)) return;
  if (!semverCheck(options.pkg.version, latestCachedVersion)) return;

  notifyFlow(options);
  if (updateAvailable) {
    return {
      updateAvailable: updateAvailable,
      name: options.pkg.name,
      latest: latestCachedVersion,
      current: options.pkg.version
    };
  }
};

module.exports = {
  updateNotify,
  pingRegistry
};
